room/compiler-processing/src/main/java/androidx/room/compiler/processing/ksp/KSTypeExt.kt

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright 2020 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package androidx.room.compiler.processing.ksp
+
+import com.squareup.javapoet.ClassName
+import com.squareup.javapoet.ParameterizedTypeName
+import com.squareup.javapoet.TypeName
+import org.jetbrains.kotlin.ksp.symbol.KSClassifierReference
+import org.jetbrains.kotlin.ksp.symbol.KSDeclaration
+import org.jetbrains.kotlin.ksp.symbol.KSName
+import org.jetbrains.kotlin.ksp.symbol.KSType
+import org.jetbrains.kotlin.ksp.symbol.KSTypeReference
+
+// catch-all type name when we cannot resolve to anything.
+private val UNDEFINED = ClassName.get("androidx.room.compiler.processing.kotlin.error", "Undefined")
+
+/**
+ * Turns a KSTypeReference into a TypeName
+ *
+ * We try to achieve this by first resolving it and iterating.
+ * If some types cannot be resolved, we do a best effort name guess from the KSTypeReference's
+ * element.
+ */
+internal fun KSTypeReference?.typeName(): TypeName {
+    if (this == null) {
+        return UNDEFINED
+    }
+    val resolved = resolve()
+    return resolved?.typeName() ?: fallbackClassName()
+}
+
+private fun KSTypeReference.fallbackClassName(): ClassName {
+    return (element as? KSClassifierReference)?.let {
+        ClassName.get("", it.referencedName())
+    } ?: UNDEFINED
+}
+
+private fun KSName.typeName(): ClassName? {
+    if (asString().isBlank()) {
+        // fallback to reference
+        return null
+    }
+    return ClassName.get(getQualifier(), getShortName())
+}
+
+private fun KSDeclaration.typeName(): ClassName? {
+    // we are only interested in qualified name as if it does not exist, it is an error for Room
+    return qualifiedName?.typeName()
```

Comment:
Does the qualified name you are looking for refers to [JVM qualified name](https://docs.oracle.com/javase/specs/jls/se11/html/jls-6.html#jls-6.7)? context on this ask: https://github.com/android/kotlin/issues/23

Reply:
actually we need package name and simple name separated, i think it is the same request as in that ticket.

I'm actually not sure if it will have a practical error in Room as I think imports will still work but would be good to make sure KSName represents the package and simple name properly (same request as in that ticket)

Reply:
In that sense it is slightly different from JVM qualified name I mentioned. Thanks for confirming.

Reply:
added notes to refer to the bug so that we can update it here once KSP has the support.

Reply:
sounds good. I've raised priority of that issue, should be able to ship before next release.